File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.py
from collections import deque

A, B = map(int,input().split())
N = 100000
dist = [N+1]*(N+1)
dist[A] = 0

doubled = [False]*(N+1)
def bfs(start):
    global doubled
    Q = deque([start])
    while Q:
        i = Q.popleft()
        if i == B:
            return dist[B]
        
        j = 2*i
        if i != 0 and not doubled[i]:
            while 0<=j<=N:
                if dist[j] == N+1:
                    dist[j] = dist[i]
                    doubled[j] = True
                    Q.append(j)
                j *= 2
        
        for j in [i-1, i+1]:
            if 0<=j<=N and dist[j] == N+1:
                dist[j] = dist[i] + 1
                Q.append(j)
        
print(bfs(A))

File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.js
const [S, E] = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "input.txt")
  .toString()
  .split(" ")
  .map(Number);
const Q = [[S, 0]];
let idx = 0;
const dist = Array(100_001).fill(Infinity);
dist[S] = 0;

while (Q.length > idx) {
  const [i, d] = Q[idx++];

  if (2 * i <= 100_000 && dist[2 * i] > d) {
    dist[2 * i] = d;
    Q[--idx] = [2 * i, d];
  }

  for (let j of [i - 1, i + 1]) {
    if (j < 0 || j > 100_000) continue;
    if (dist[j] <= d + 1) continue;
    dist[j] = d + 1;
    Q.push([j, d + 1]);
  }
}

console.log(dist[E]);
